readme.md
## Norwich: A Kind City
### Instaltion and Build

Fork the Repo from GitHub.
The latest version is in the **master** branch, and the latest deployment in **gh-pages** branch

Install dependencies with:
`````
npm install
npm install -g parcel-bundler
`````

For a local sever:
`````npm start`````

To Deploy to Github.io:
`````npm run deploy`````

The demo page is: https://robinkeith.github.io/AKindCity6

### Build Process
The project does not have a sever-side component - at runtime everything is contained within a single client-side page (and a service worker). Maptiles are provided by Mapbox and points of interest data is provided by static files of data. 

There is an offline process to generate the data files from extracts of OSM data, which is processed to improve readability. The offline data import is contained in the /backend directory, and run using Debug > Import. The import process runs in a local nodejs, long term this will be run automatically each day/week on a server.  Data files are generated in the /data directory. The process also generates some statistics files in /data/csv and uses /data/temp to store files during processing.

The build process uses ParcelJS, which is a bundler similar to Webpack.  It is largely auto configuring, although there are couple of plug-ins required (and configured by package.json). ParcelJS uses Bable unde the hood to transpile for the browser.
The code uses ES module notation for imports (eg. import * from XXXX) rather than commonJS (ie. var xxx=require(XXXX)) which will transpiled as needed. In theory this should minimise the volume of code in the final bundle....
Resolving incompatabilities between libraries using the two import methods has been one of the biggest challenges in the project - especially when sharing between the from and backends, the openinghours module and Leaflet. Leaflet still appears to be including everything, and needs further work to determine why.
Although npm libraries have been used for the most part, there are some modules which are pulled directly from github forks. Opening hours was particually tricky as npm does yet contain a version which includes uk PH dates.
Leaflet.vectoricons - the npm version is pinned to leaflet 0.7

Leaflet's plug-in mechanism is a bit tricky to use with ES modules. Plug-ins need to have been tweaked to use the universal module loader code mentioned on the leaflet site. Some projects have, but not all. Also note subtle changes in the project names (e.g. caps changes and . changed to -) are significant. The project uses jsdoc as a step towards full typescript, and the plug-ins tend not to play well with that.

The build and deploy processes include copying all files in the **data** folder to the **dist** folder
The build process also clears the build cache and dist folder. This is to avoid occasional problems with the build failing due to cache issues. Although this slows the build, it's not too noticeable as most changes are hot reloadable.

The following can be ignored:
 WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@
  It occurs because fsevents is an optional dependency, used only when project is run on macOS environment (the package provides 'Native Access to Mac OS-X FSEvents'). And since you're running your project on Windows, fsevents is skipped as irrelevant.

  ## Favicon and App icons
  Generated using https://realfavicongenerator.net/
  using the resources/tcsm_pointer.svg 

  The process generates a zip file of icons. Unip them to the root of the project folders.
  
  ## Data Import 
  Contained in the /back-end/ folder. 
  Run Debug task 'download' to refresh data from OSM.

  
 ## PWA Functionality
 Plan:
 * Create a Service Worker
 * Mobile first layout
 * Offline map tiles

 ### Service worker

Generate a Complete Service Worker with Workbox CLI https://developers.google.com/web/tools/workbox/guides/generate-service-worker/cli

npm install workbox-cli --save-dev
workbox wizard

Uses parcel-plugin-workbox to inject workbox into the seup. parcel-plugin-sw was oiginally used, but no longer.
parcel-plugin-workbox uses the workbox property in package.json, but we use the defaults.
`````
"workbox": {
  importScripts: ['./worker.js'],           // scripts to import into `sw.js`
  globDirectory: './dist',                  // directory to cache (usually output dir)
  globPatterns: [                           // file types to include
  '**/*.{css,html,gif,js,jpg,png,svg,webp}'
}
`````

For reference, the plugin-sw config which is no longer used
"commented_cache - not used": {
    "inDev": true,
    "strategy": "inject",
    "swSrc": "src/service-worker-templates.js",
    "swDest": "map-service-worker.js"
  },

### Potential candidates for other layer styles
 https://openmaptiles.github.io/klokantech-3d-gl-style/#17/52.62690/1.29217/-42.5/57 
 https://openmaptiles.org/styles/
 

## Opening Hours
The OSM opening hours format is quite complex.  The npm package opening_hours is comprehensive,
but out of date, so we use the git repo instread
`````
npm i github:opening-hours/opening_hours.js.git#master
````

npm cache clean --force


changelog.md
# Change Log

# Version 2
## Version 2.0
Goals: Improve presentation of map features
* Improved opening hours, easier to read, expands abbreviations. Accomodates new OSM tags for kitchen hours etc
* Moved 'currently open/closed' to right hand side and made more prominent
* Improved display of location/address

TODO:
1. Generate description in postprocessing. Replace current list of tags with readable descriptions, customised for the type of attribute
2. Include safe places and blue badge in postprocessin
3. Add hover captions, improve when to display popup vs hover
4. Hide website, phone + email links with icoms
5. Introduce specific icons for shop, help and enjoy types
6. Allow filtering of layer by type
7. Highlight the 'nearest' when tracking is on.
8. Dim currently closed features, part dim closing soon features
9. Time travel - move map back and forward in time to show when things open.


## 1.9
* Improvements to 'get around' layer - colour coding added - red for obsticles (stairs), green for aids (lifts, ramps, tactile surfaces etc), amber for impediments (e.g. steep inclines)
* Added feedback form - press the 'speech bubble icon' links to the online surveys
* Added feedback button to popup window. The OSM ref is copied to the clipboard. Updated suveys to prompt for the reference
* fixed a bug in personalisation screen and loading one of the toilet layers
* fixed bug where caption showed previous POI if the new point had no caption.
* Data preprocessing colapses outlines to points automatically
* Moved polygon to point conversion to overpass query - reduces layer loading time.
* Move name tidy to preprocessing to reduce layer load time.
* Improved toolbar display on laptops and larger screen sizes
* Implemented basic service worker for PWA compatability. Map data is cached off-line. Map tiles visited will be for offline use but not yet downloading all tiles
* Opening Hours improvements, shows if places are currently open or close and next opening/closing time.
* Improvements to OSM data, added opening hours, website + email details, fixed out of date tags, and added 3d info for large number of city centre places.
Known issues: The 'Get Around' layer data for crossings etc is missing


## 1.8
* Added Elephant Juice! Press the elephant button to remember where you parked.
* Blue badge parking info - imported but not on preview website due to licencing restrictions (being worked through with NCC)
* Reorganised Parking icons to show parking meters and large car parks more clearly.
* As requested by NCC, the SafePlaces data now requires a pass code - Tick the 'Sometimes I need a little more help' option in 'All About Me' and enter the pass code.
* Product tour - the first time the app is used you get a quick walk-through of how to use it. Repeat the tour from the 'Take the Tour' button on the Personalization forrm.


## 1.7
* Added layers for 'shop' and 'enjoy'
* Added get around layer (ramps, steps, crossings, lifts, escalators)
* Added shop mobility to the 'around' layer
* increased size of top left icons


## 1.6
* Scripted import from OSM (still on demand, but single button to refresh the data) 
* Added Changing Places data and approved logo
* updated favicon + app icons to new design
* Tweaks to layout
* fix: user settings not refreshed after saving.
* Marker feature change to modal box and made larger
* Layer buttons only shown if layer contains data (the other layers will return shortly!)
* refactoring of code around popups and layer loading - popup now take whole screen and much large text

## 1.5
* Link to open street map editor fixed
* Improved popup information
* changed icon for parking meter
* Added Clare School Logo
* Formatted Layer selector
* added feature search
* groupped buttons into top left (on screen controls temp. removed)

## 1.4
* Settings for wheelchair and limited mobility changed to switch buttons. Reworked filter functions to account
* improved saving of personalisation settings to localstorage
* collated defaults into one place
* changed map bounds and max/min zoom to suite
* added edge buffering for smoother map
* improved peformance of layer loading to <1s
* popup - defaults to supplied by OSM with link to edit the feature in OSM.

## 1.3
* Improvements to toilet layer, start of common method for massarging OSM data into a more consistent form.
* Impoved popup display for toilets amenity layer.
* Personalisation iteration 1. layers can filter data based on wheelchair via icon bottom right
* Added EAT layer
* Active layers are saved to localStorage and recalled next session
* Opening_hours parsing for current status in place but needs more work.


## 1.2
* Map position and zoom saved for next time
* Updated the title to 'The Clare School Map'
* Toilets are always rendered as markers, even when in buildings
* Improved feature descriptions - click markers on the toilets layer. (much more to do here!)
* Added one of the Changes Places (Castle Mall) to OSM
* Added the pedestrian area outside The Forum, and various other fixes to OSM


## 1.1
* Improved toilet data to include toilets in buildings (when listed in OSM)
* Added Full Screen button
* Added Geolocate (you are here) button 
* Added button for personalisation (non-functional)
* Added basic routing control - very basic as yet
* Icons - a little icon will appear in the top of the browser and if you add it to your home page on an iPad/andoid.


demo/demo.js
// “Mapbox GL JS is a JavaScript library that uses WebGL to render interactive maps from vector tiles and Mapbox styles.” -- https://www.mapbox.com/mapbox-gl-js/api/
mapboxgl.accessToken = 'pk.eyJ1IjoidXgiLCJhIjoiY2l2Zm5xZmE2MDAzOTJvcGZ4dngwbWw2aSJ9.XhTQ80_U9ZWsoqzqxHDEwQ'

if (!mapboxgl.supported()) { // https://www.mapbox.com/mapbox-gl-js/plugins/#mapbox-gl-supported
  alert('Your browser does not support Mapbox GL')
} else {
  var mbStyleStr8 = 'mapbox://styles/mapbox/streets-v8' // Streets v8
  var mbStyleStr9 = 'mapbox://styles/mapbox/streets-v9'
  var mbStyleOutd = 'mapbox://styles/mapbox/outdoors-v9'
  var mbStyleLight = 'mapbox://styles/mapbox/light-v9'
  var mbStyleDark = 'mapbox://styles/mapbox/dark-v9'
  var mbStyleBright = 'mapbox://styles/mapbox/bright-v9'

  var lnglatTallinn = [24.753574, 59.436960] // Tallinn
  var lnglatTallinnTvtower = [24.8875, 59.471206] // Tallinn TV Tower
  var lnglatTallinnTownhall = [24.745319, 59.437372] // Tallinn Town Hall
  var lnglatTallinnTowers = [24.757888, 59.433184] // Tallinn city
  var lnglatTallinnSkype = [24.661038, 59.397467] // near Skype office in Tallinn

  var angleTallinnTownhall = {
    center: lnglatTallinnTownhall,
    zoom: 17.5, // 15 is the min. zoom level showing the whole 3D buildings layer
    pitch: 60, // pitch (tilt) of the map, measured in degrees away from the plane of the screen (0-60)
    bearing: -150 // bearing (rotation) of the map, measured in degrees counter-clockwise from north
  }
  var angleTallinnTowers = {
    center: lnglatTallinnTowers,
    zoom: 15.5,
    pitch: 55,
    bearing: -50
  }
  var angleTallinnTvtower = {
    center: lnglatTallinnTvtower,
    zoom: 16,
    pitch: 30,
    bearing: -220
  }
  var angleTallinnSkype = {
    center: lnglatTallinnSkype,
    zoom: 17,
    pitch: 30,
    bearing: -225
  }

  var map = new mapboxgl.Map({

    container: 'map',
    style: mbStyleStr9,

    center: angleTallinnTownhall.center,
    zoom: angleTallinnTownhall.zoom,
    pitch: angleTallinnTownhall.pitch,
    bearing: angleTallinnTownhall.bearing

  })

  var currentLng, currentLat, currentZoom, currentPitch, currentBearing

  var logMapProps = function () {
    var centerObj = map.getCenter()
    currentLng = centerObj['lng']
    currentLat = centerObj['lat']

    currentZoom = map.getZoom()
    currentPitch = map.getPitch()
    currentBearing = map.getBearing()

    $('#log-lng').html(parseFloat(currentLng.toFixed(6)))
    $('#log-lat').html(parseFloat(currentLat.toFixed(6)))
    $('#log-zoom').html(parseFloat(currentZoom.toFixed(6)))
    $('#log-pitch').html(parseFloat(currentPitch.toFixed(6)))
    $('#log-bearing').html(parseFloat(currentBearing.toFixed(6)))
  }

  var flyStart = angleTallinnTowers

  var $flyBtn = $('#fly')

  map.on('load', function () {
    map.addLayer({
      'id': '3d-buildings',
      'source': 'composite',
      'source-layer': 'building', // building-height data from OpenStreetMap
      'filter': ['==', 'extrude', 'true'],
      'type': 'fill-extrusion',
      // 'minzoom': 15,
      'paint': {
        'fill-extrusion-color': '#aaa', // #aaa === 33% black
        'fill-extrusion-height': {
          'type': 'identity',
          'property': 'height'
        },
        'fill-extrusion-base': {
          'type': 'identity',
          'property': 'min_height'
        },
        'fill-extrusion-opacity': 0.5
      }
    })

    logMapProps()

    setTimeout(function () {
      map.flyTo({ // https://www.mapbox.com/mapbox-gl-js/api/#Map#flyTo

        center: flyStart.center,
        zoom: flyStart.zoom,
        pitch: flyStart.pitch,
        bearing: flyStart.bearing,

        speed: 0.65, // make the flying slow (default: 1.2)
        curve: 0.95, // change the speed at which it zooms out (default: 1.42)

        easing: function (t) {
          return t
        }

      })

      setTimeout(function () { $flyBtn.show() }, 4000)
    }, 1500)
  })

  var isAtStart = true
  var nextIsSkype = true
  var flyTarget

  $flyBtn.on('click', function () {
    // Switch b/w points A/B/C, and set the corresponding flags
    isAtStart ? (
      flyTarget = nextIsSkype ? angleTallinnSkype : angleTallinnTvtower,
      nextIsSkype = !nextIsSkype,
      isAtStart = !isAtStart
    ) : (
      flyTarget = flyStart,
      isAtStart = !isAtStart
    )

    map.flyTo({

      center: flyTarget.center,
      zoom: flyTarget.zoom,
      pitch: flyTarget.pitch,
      bearing: flyTarget.bearing,

      speed: 0.35,
      curve: 0.7,

      easing: function (t) {
        return t
      }

    })
  })

  map.on('move', function () {
    logMapProps()
  })
}
